### PR TITLE
feat(shoplist): stabilize plan->shoplist flow package

### PR DESCRIPTION
**Pull request by Katsiarynakavaleuskaya**
## Summary
- stabilize the canonical `POST /api/v1/pro/meal/shopping-list` flow to produce deterministic plan-to-shoplist output ordering
- harden extraction/generation against nondeterministic input ordering and invalid quantity shapes without adding new runtime surfaces
- add focused runtime coverage: one deterministic happy-path integration test and one negative matrix for auth/validation plus no-leak behavior on unexpected errors

## Test plan
- [x] `pytest -q tests/test_shoplist_flow_integration.py tests/test_shoplist_flow_negative_matrix.py tests/test_shopping_list_pro.py tests/test_shopping_list_pro_router.py tests/core/test_shoplist_contract.py`
- [x] `pre-commit run --all-files`
- [x] `make verify`

## Scope freeze
- IN: `app/core/shopping_list/*`, `core/shoplist.py`, flow-focused tests under `tests/`
- OUT: docs-only updates, iOS/web changes, CI/workflow changes, security suppression updates

## Deferred / Follow-ups
- None.

## Discussion Thread Pass
- [x] Discussion-thread pass completed
- [x] Fixed in commit mapping completed

### Fixed in Commit Mapping
- No actionable review comments

Made with [Cursor](https://cursor.com)

## Summary by Sourcery

Стабилизировать процесс преобразования плана в список покупок, чтобы получать детерминированный, устойчивый результат, а также усилить обработку некорректных количественных значений.

Исправления ошибок:
- Игнорировать ключи ингредиентов, которые не являются строками, и некорректные структуры количеств при извлечении ингредиентов и оптимизации упаковки.
- Гарантировать, что агрегированные категории и элементы списка покупок генерируются в детерминированном, отсортированном порядке.

Тесты:
- Добавить интеграционный тест, проверяющий детерминированные ответы преобразования плана в список покупок для корректного ввода.
- Добавить матрицу негативных тестов, охватывающую ошибки аутентификации/валидации и единообразные форматы ошибок для pro-эндпоинта списка покупок.
- Добавить тест, гарантирующий, что неожиданные серверные ошибки в процессе формирования списка покупок не раскрывают внутренние детали или секреты в ответах.

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Stabilize the plan-to-shopping-list flow to produce deterministic, robust output and harden handling of malformed quantities.

Bug Fixes:
- Ignore non-string ingredient keys and invalid quantity shapes when extracting ingredients and optimizing packaging.
- Ensure aggregated shopping-list categories and items are generated in a deterministic, sorted order.

Tests:
- Add an integration test verifying deterministic plan-to-shopping-list responses for valid input.
- Add a negative test matrix covering auth/validation failures and consistent error envelopes for the pro shopping-list endpoint.
- Add a test ensuring unexpected server errors on the shopping-list flow do not leak internal details or secrets in responses.

</details>

---
## Summary by cubic
Stabilizes the plan-to-shopping-list flow so identical inputs produce deterministic, sorted output, regardless of meal or ingredient order. Broadens and type-hardens quantity parsing, and normalizes meal titles.

- **Bug Fixes**
  - Enforce deterministic ordering by sorting ingredient keys during extraction and iterating aggregated keys in order; normalize meal titles (trim and fallback key).
  - Coerce quantities via float() with explicit accepted types (str/bytes/SupportsFloat/SupportsIndex), removing Any and keeping broad numeric-protocol support; skip only when conversion fails.
  - Add focused tests: end-to-end determinism (including reordered inputs), a negative auth/validation matrix, and a no-leak 500 check.

- **Refactors**
  - Simplify optimize_packaging quantity coercion and remove a dead continue branch; no behavior change.

<sup>Written for commit cfa036e2d60d84851e6f5e6e41875ef3fd4b34ca. Summary will update on new commits.</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Shopping lists now generate deterministically with stable, sorted categories and items; recipe references normalize meal titles (trim/empty-title handling).
  * Invalid or malformed quantity values are ignored; packaging accepts a broader set of numeric-like inputs.
  * Error responses no longer expose internal tracebacks or file-path details.

* **Tests**
  * Added integration and negative tests to verify deterministic ordering, stable output, filtering of bad quantities, and secure error handling.

### CODE REVIEW

**Comment by Katsiarynakavaleuskaya** [comment]
Addressed latest bot actionables with focused fixes.

## Fixed in commits
- `ea81a34f` — core quantity coercion now relies on `float()` + exception handling (accepts Decimal/numeric-protocol inputs; no narrowing guard).
- `81072e6f` — test updates:
  - determinism integration now validates reordered input (meals + grams key order)
  - negative matrix is status-only (guard consistency)
  - no-leak test now uses shared `app` fixture instead of direct `TestClient(app.main.app)`.

## Validation
- `pytest -q tests/test_shoplist_flow_negative_matrix.py tests/test_shoplist_flow_integration.py tests/core/test_shoplist_contract.py` ✅
- `pre-commit run --all-files` ✅
- `make verify` ✅

Thanks for the reviews @sourcery-ai @coderabbitai @cubic-dev-ai.

**Comment by Katsiarynakavaleuskaya** [comment]
Addressed the latest Cubic actionable.

- Fixed in `e4483cb0`: replaced the core-local `Any` usage with explicit `object` + typed coercion cast for `float(...)`, preserving runtime behavior while keeping core typing strict.

Validation after the change:
- `pre-commit run --all-files` ✅
- `make verify` ✅

@cubic-dev-ai please re-run review.

File: tests/test_shoplist_flow_integration.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+def test_plan_to_shoplist_flow_is_deterministic(
+    client: TestClient, pro_headers: dict[str, str]
+) -> None:
+    """Same plan input must produce stable shoplist output."""
+    plan_data = {
+        "daily_menus": [
+            {
+                "meals": [
+                    {
+                        "title": "late_meal",
```
[the remark this answers is not in the document]
Comment (Katsiarynakavaleuskaya) [reply tests/test_shoplist_flow_integration.py:24]: Addressed in `81072e6f`: added reordered-input determinism variant (reversed meals and reversed grams insertion order) and assert output equality via `_without_generated_at(...)` against baseline.

**Comment by Katsiarynakavaleuskaya** [comment]
Addressed latest CodeRabbit nitpicks in : removed redundant  after  and simplified quantity coercion by casting  directly (no temporary alias). Local and CI checks are green; thread can be considered resolved.
